bootstrap/cmd/bootstrap/app/kfctlServer.go

```diff
@@ -60,15 +301,214 @@ func (r *kfctlServer) RegisterEndpoints() {
 	}
+
+	if new == nil {
+		return false
+	}
+
+	if current.Name == "" && current.Spec.Project == "" && current.Spec.Zone == "" {
+		return true
+	}
+
+	if current.Spec.Project != new.Spec.Project {
+		return false
+	}
+
+	if current.Spec.Zone != new.Spec.Zone {
+		return false
+	}
+
+	if current.Name != new.Name {
+		return false
+	}
+
+	return true
+}
+
+// prepareSecrets prepares the secrets in the request.
+// Literal secrets are converted to environment variables except for the GcpAccessToken which is removed.
+//
+// TODO(https://github.com/kubeflow/kubeflow/issues/3592) Once the apply methods take a context we should be able
+// to use that and not rely on the environment
+func prepareSecrets(d *kfdefsv2.KfDef) {
+	secrets := []kfdefsv2.Secret{}
+
+	for _, s := range d.Spec.Secrets {
+		// Don't pass along the access token
+		if s.Name == gcp.GcpAccessTokenName {
+			continue
+		}
+
+		if s.SecretSource.LiteralSource != nil {
+			n := "KFCTL_" + s.Name
+			s.SecretSource.EnvSource = &kfdefsv2.EnvSource{
+				Name: n,
+			}
+
+			os.Setenv(n, s.SecretSource.LiteralSource.Value)
+			s.SecretSource.LiteralSource = nil
+
+		}
+
+		secrets = append(secrets, s)
+	}
+
+	d.Spec.Secrets = secrets
+}
+
+// CreateDeployment creates the deployment.
+//
+// Not thread safe
+// TODO(jlewi): We should check if the request matches the current deployment and if not reject
+func (s *kfctlServer) CreateDeployment(ctx context.Context, req kfdefsv2.KfDef) (*kfdefsv2.KfDef, error) {
+	token, err := req.GetSecret(gcp.GcpAccessTokenName)
+
+	if err != nil {
+		log.Errorf("Failed to get secret %v; error %v", gcp.GcpAccessTokenName, err)
+		return nil, &httpError{
+			Message: fmt.Sprintf("Could not obtain an access token from secret %v", gcp.GcpAccessTokenName),
+			Code:    http.StatusBadRequest,
+		}
+	}
+
+	initFunc := func() error {
+		s.kfDefMux.Lock()
+		defer s.kfDefMux.Unlock()
+
+		if s.ts == nil {
+			log.Infof("Initializing token source")
+			ts, err := NewRefreshableTokenSource(req.Spec.Project)
+
+			if err != nil {
+				log.Errorf("Could not create token source; error %v", err)
+
+				return &httpError{
+					Message: "Internal service error please try again later.",
+					Code:    http.StatusInternalServerError,
+				}
+			}
+
+			s.ts = ts
+		}
+		return nil
+	}
+
+	if err := initFunc(); err != nil {
+		return nil, err
+	}
+
+	// Refresh the credential. This will fail if it doesn't provide access to the project
+	err = s.ts.Refresh(oauth2.Token{
+		AccessToken: token,
+	})
+
+	if err != nil {
+		log.Errorf("Refreshing the token failed; %v", err)
+		return nil, &httpError{
+			Message: fmt.Sprintf("Could not verify you have admin priveleges on project %v; please check that the project is correct and you have admin priveleges", req.Spec.Project),
+			Code:    http.StatusBadRequest,
+		}
+	}
+
+	checkIsMatch := func() bool {
+		s.kfDefMux.Lock()
+		defer s.kfDefMux.Unlock()
+		return isMatch(&s.latestKfDef, &req)
+	}
+
+	if !checkIsMatch() {
```

Comment:
If `checkIsMatch == true`, we don't need to push same request to channel?

Reply:
We do down down below; line 466. If the request is a match then we keep processing.
We check that it is a valid request and if it is we keep going which means we add it to the channel to actually get processed.